sdk/python/feast/infra/registry/registry.py

```diff
@@ -338,10 +338,27 @@ def list_data_sources(
     def apply_data_source(
         self, data_source: DataSource, project: str, commit: bool = True
     ):
+        now = _utc_now()
+        if not data_source.created_timestamp:
+            data_source.created_timestamp = now
+        data_source.last_updated_timestamp = now
+
         registry = self._prepare_registry_for_changes(project)
+
         for idx, existing_data_source_proto in enumerate(registry.data_sources):
             if existing_data_source_proto.name == data_source.name:
-                del registry.data_sources[idx]
+                existing_data_source = DataSource.from_proto(existing_data_source_proto)
+                # Check if the data source has actually changed
+                if existing_data_source == data_source:
+                    return
+                else:
+                    # Preserve created_timestamp from existing data source
+                    data_source.created_timestamp = (
+                        existing_data_source.created_timestamp
+                    )
```

Comment:
Pritty strange why we want to preserve the timestamp of datasource which we are about to delete ?

Reply:
The logic is not about deletion. It finds existing data source in registry, If it's different from the new one, preserve the original created_timestamp (because the data source was first created at that time), delete the old version from registry, add the new version with the preserved created_timestamp. This ensures created_timestamp represents when the data source was first created, not when this particular update happened.